### DISCUSSION TITLE
MeshCache design for devices

### DISCUSSION THREAD

**Discussion by ecoon** [amanzi/amanzi#675] [General]
### How to write users of MeshCache to not constrain MeshCache

This is to help me organize my thoughts in a place where I can format code, and hopefully explain to @JulienLoiseau some design requirements.  Comments welcome.

Note here I will use the term "cached" to mean that the MeshCache object stores memory for this object that is precomputed, and "computed" to mean that the MeshCache object does not store memory.

Consider the following user of MeshCache (from PDE_DiffusionFV, tpetra branch, with some updates for newer naming):  

```
// VERSION 1
const MeshCache<DEVICE> m;
Kokos::View<double*, DEVICE> trans_face; // this is a device view from a Tpetra::Vector

Kokkos::parallel_for(
        "PDE_DiffusionFV::ComputeTransmissibility",
        ncells_owned,
        KOKKOS_LAMBDA(const int c) {
          Kokkos::View<Entity_ID*, DEVICE> faces;
          Kokkos::View<AmanziGeometry::Point*, DEVICE> bisectors;
          std::tie(faces, bisectors) = m->getCellFacesAndBisectors(c);

          WhetStone::Tensor<DEVICE> Kc = K->at(c);
          
          for (int i = 0; i < faces.extent(0); i++) {
            auto f = faces(i);
            const AmanziGeometry::Point& a = bisectors(i);
            const AmanziGeometry::Point& normal = m->getFaceNormal(f);
            const double area = m->getFaceArea(f);

            const double h_tmp = AmanziGeometry::norm(a);
            const double s = area / h_tmp;
            const double perm = ((Kc * a) * normal) * s;
            const double dxn = a * normal;
            Kokkos::atomic_add(&trans_face(f,0), fabs(dxn / perm));
          }
        });
```

Ideally, this would actually be written via hierarchical parallelism, but we skip that for now because you can't capture the atomic inside the inner loop, and we can't invert the loop (e.g. looping over faces and doing a parallel reduce for each face) because that would be an expensive search and weird memory access.

But also notice that this presupposes a lot:

1. We get a reference to a normal.  This implies that the normal was cached.  There is no way to compute a normal and return the temporary.
2. We get Kc by value.  Kc could have either been cached or computed, but if it was cached, we made a copy (in this case, Kc is a shallow copy view).
3. We create (sub)views for both faces and bisectors.  This implies that these quantities were either computed or cached, but if they were computed, a malloc was required to store the view.

## Universal References

The combination of 1 & 2 demonstrate the first conundrum -- capture cached things by reference and hard-code in semantic, vs capture all things by value and hard-code in an unnecessary copy construction.

First, let's try to fix 1 and 2: In an ideal world, the user would not know whether things are cached or computed.  Maybe we can get around this by using universal references?  E.g. should we have written:

```
// VERSION 2
const MeshCache<DEVICE> m;
Kokos::View<double*, DEVICE> trans_face; // this is a device view from a Tpetra::Vector

Kokkos::parallel_for(
        "PDE_DiffusionFV::ComputeTransmissibility",
        ncells_owned,
        KOKKOS_LAMBDA(const int c) {
          Kokkos::View<Entity_ID*, DEVICE> faces;
          Kokkos::View<AmanziGeometry::Point*, DEVICE> bisectors;
          std::tie(faces, bisectors) = m->getCellFacesAndBisectors(c);

          WhetStone::Tensor<DEVICE>&& Kc = K->at(c);
          
          for (int i = 0; i < faces.extent(0); i++) {
            auto f = faces(i);
            const AmanziGeometry::Point& a = bisectors(i);
            const AmanziGeometry::Point&& normal = m->getFaceNormal(f);
            const double area = m->getFaceArea(f);

            const double h_tmp = AmanziGeometry::norm(a);
            const double s = area / h_tmp;
            const double perm = ((Kc * a) * normal) * s;
            const double dxn = a * normal;
            Kokkos::atomic_add(&trans_face(f,0), fabs(dxn / perm));
          }
        });
```

Note the difference in Kc and normal -- both are captured (one by value, one by reference, assuming no upstream code has changed) by the same universal reference.  In particular, if the MeshCache was changed to compute the face normal and return by **value** to the temporary, the kernel would not have to change.

@JulienLoiseau Is this CUDA safe?  Are universal references ok in CUDA?  Is there another reason to not do this?

## Avoiding subview construction
Next let's look at 3.  Can we avoid making the bisector and faces subviews?

```
// VERSION 3
const MeshCache<DEVICE> m;
Kokos::View<double*, DEVICE> trans_face; // this is a device view from a Tpetra::Vector

Kokkos::parallel_for(
        "PDE_DiffusionFV::ComputeTransmissibility",
        ncells_owned,
        KOKKOS_LAMBDA(const int c) {
          WhetStone::Tensor<DEVICE>&& Kc = K->at(c);
          
          auto c_nfaces = m.getCellNumFaces(c);
          for (int i = 0; i != c_nfaces; ++i) {
            auto f = m.getCellFace(c,i);
            auto&& a = m.getCellFaceBisector(c,i);
            auto&& normal = m->getFaceNormal(f);
            const double area = m->getFaceArea(f);

            const double h_tmp = AmanziGeometry::norm(a);
            const double s = area / h_tmp;
            const double perm = ((Kc * a) * normal) * s;
            const double dxn = a * normal;
            Kokkos::atomic_add(&trans_face(f,0), fabs(dxn / perm));
          }
        });
```

Now, if the bisectors are computed, we can actually do them on the fly rather than mallocing a view, populating the view, then using the view.  I think this is better in all cases -- in the case of cached bisectors, we have avoided creating a subview, and in the case of computed bisectors, we have avoided having to malloc the view.

### Can MeshCache cache or compute?

So this brings us to Algorithms.  Note that in the above usage, we implied that we could compute things one-at-a-time.  For instance, if we compute face normals, `getFaceNormal(int f)` must ONLY compute f, not "all the normals".  In fact, getFaceNormal(f) gets called twice for each internal face but only once for each external face.  So the algorithm must be written not for a full view, but for a single entry in order to be useful here.

The current prototype only computes algorithms as "on the full view."  This implies a pattern of "cache just before you use the cache, then use the cache, then throw away the cache," which isn't computing, it's just caching with a different lifespan.  In that pattern, we should have written:

```
// VERSION 4
const MeshCache<DEVICE> m;
Kokos::View<double*, DEVICE> trans_face; // this is a device view from a Tpetra::Vector

m.cacheFaceNormals();
m.cacheFaceBisectors();

Kokkos::parallel_for(
        "PDE_DiffusionFV::ComputeTransmissibility",
        ncells_owned,
        KOKKOS_LAMBDA(const int c) {
          WhetStone::Tensor<DEVICE>&& Kc = K->at(c);
          
          auto c_nfaces = m.getCellNumFaces(c);
          for (int i = 0; i != c_nfaces; ++i) {
            auto f = m.getCellFace(c,i);
            auto& a = m.getCellFaceBisector(c,i);
            auto& normal = m->getFaceNormal(f);
            const double area = m->getFaceArea(f);

            const double h_tmp = AmanziGeometry::norm(a);
            const double s = area / h_tmp;
            const double perm = ((Kc * a) * normal) * s;
            const double dxn = a * normal;
            Kokkos::atomic_add(&trans_face(f,0), fabs(dxn / perm));
          }
        });

m.destroyFaceNormals();
m.destroyFaceBisectors();

```

or something like that.  Note that this would be potentially 5 kernel launches: one to compute face normals, one to compute face bisectors, two (implicitly called within the bisectors routine) for cell_centroids and face_centroids, and finally the outer kernel.  

So I think that isn't that helpful.  The whole point of "computing" is that some of these things are relatively cheap to compute, and there are a LOT of them, so caching everything would blow up our memory footprint.  Sure, we could cache and uncache each time we use something, but it would be better if these "algorithms" would be non-collective, KOKKOS_INLINE_FUNCTIONS that could be called either on host or device.  Think things like:

```
// computes the ith face of cell c's bisector
template<typename Mesh_type>
AmanziGeometry::Point computeBisector(const Mesh_type& m, Entity_ID c, Entity_ID i) {
  return m.getFaceCentroid(m.getCellFace(c,i)) - m.getCellCentroid(c);
}
```

Can this be a KOKKOS_INLINE_FUNCTION, and be used by both MeshCache (on host or device) and MeshFramework (on host only), without having to maintain multiple copies (e.g. in the current prototype, where algorithms have multiple copies that only differ by the loop "invocation")?

### Function pointers

Lastly, presume that at runtime, depending upon the exact physics called, we decide to cache bisectors rather than recompute them because we use them all the time.

Then, we can call `m.cacheBisectors();` which does a parallel for, calling the algorithm, and storing the result.  Can this function call would then swap the function pointers, between two implementations:

```
template<typename MEM>
KOKKOS_INLINE_FUNCTION
const AmanziGeometry::Point&
MeshCache<MEM>::getCellFaceBisector(Entity_ID c, int i) const {
  return bisectors_.get<MEM>(c,i); // note, bisectors_ is a CSR, so I'm presuming a 
                                  // get method that is templated on device to get the right 
                                  // view in the dual-view CSR.
}

template<typename MEM>
KOKKOS_INLINE_FUNCTION
AmanziGeometry::Point
MeshCache<MEM>::getCellFaceBisector(Entity_ID c, int i) const {
    return Algorithms::computeCellFaceBisector(*this, c, i);
}
```

Note that the return type is different!  The first one returns a const reference while the second returns a temporary by value.  Does this break the pointer swap?  Does `auto` return help us here? (edit: no, I don't think so).

If we cannot do the function pointer swap because of differing return types, we would likely have a conditional, e.g.

```
template<typename MEM>
KOKKOS_INLINE_FUNCTION
AmanziGeometry::Point // <---- cannot put && here, sadface
MeshCache<MEM>::getCellFaceBisector(Entity_ID c, int i) const {
    if (bisectors_cached) return bisectors_.get<MEM>(c,i);
    else return Algorithms::computeCellFaceBisector(*this, c, i);
}
```

Two observations about this -- 1, Point must now be returned by value even in the cached case (bad) and 2, we now have a conditional, though the value of that conditional is guaranteed to be the same on all threads of a parallel_for if that helps.  Is that conditional branch predicted?  Does a ternary operator change things?

The accepted answer here: https://stackoverflow.com/questions/32001289/for-cuda-is-there-a-guarantee-that-ternary-operator-can-avoid-branch-divergence  suggests that the conditional won't actually matter, as CUDA will probably just do the right thing, but it still means an unnecessary copy is made.  Maybe there is no way around this.

### Sometimes we cannot compute

Lastly, there is one corner case which nevertheless has to be mentioned.  We have two main types of meshes -- unstructured meshes which use the default MeshFrameworkAlgorithms implementation, and "logical" meshes which do not have nodes, but instead provide all the geometry, or are so simple that the geometry is simply stated (like 1D meshes, or even 1-cell meshes).  For instance, in logical meshes, there is no "nodal coordinates" from which to compute face centroids and therefore no way of computing bisectors, but the user provides an array of all bisectors.

In this case, the only "algorithm" is to consult an array in the MeshFramework, so caching is the only right thing to do if you want to avoid doing the deep_copy from host to device every time.  In that case, even if the user asked us to throw away the cache, we should probably say no and keep the cache, because otherwise the next time you try to do that operation, the "compute" method would get called, which would throw an error.

This is why MeshFrameworkAlgorithms are currently virtual -- the "bisectors" algorithm for a logical mesh overrides the default algorithm to just ask the framework for the value.  Maybe there is a better way to do this as well?

**Comment by ecoon** [comment]
Conclusions (I think?):

1. We need to write algorithms as local not collective operations.
2. Maybe I am over-optimizing and we should start by not worry about return semantics and just assume everything is returned by value?  Likely even Points are small, so maybe worrying about the return-by-reference case isn't so big a deal?  This can be measured now!
3. If 2 is true and we can return by value, we may be able to also not worry about function pointers, and instead just include a conditional, as we can guarantee that all threads have the same branch, so there should never be branch divergence.  It sounds like CUDA will do the right thing, but we can check this.
4. We probably still need to figure out non-virtual algorithms somehow to avoid vtable lookups on device every time you try to compute.

**Comment by JulienLoiseau** [comment]
From our last Wednesday discussion: 

```
template<memory M> 


const Point_View
KOKKOS_INLINE_FUNCTION MeshCache::getEdgeCoordinates(const Entity_ID n) const
{
  if(edge_coordinates_cached) return getEdgeCoordinates(n); 
  return MeshAlgorithms::getEdgeCoordinates(*this, n);
}

const Point_View&
KOKKOS_INLINE_FUNCTION MeshCache::edge_coordinates(const Entity_ID n) const
{
  assert(edge_coordinates_cached); 
  return view<M>(edge_coordinates_)(n);  
}


// User 
Mesh<device>* m;

Kokkos::parallel_for("",size,KOKKOS_LAMBDA(const int c){     
  // Algo or Cached
  auto&& v = m->getEdgeCoordinates(c); 
  // Cached
  auto& v = m->edge_coordinates(c); 
});
```

**Comment by JulienLoiseau** [comment]
This doesnt solve the problem of switch from cache/algo at runtime. 
But if we want to avoid having two different functions names we can write:

```
template<storage ST = algorithm>
KOKKOS_INLINE_FUNCTION 
decltype(auto) getEdgeCoordinates(const int i) {
    if constexpr(ST == cached){
        assert(edge_coordinates_cached); 
        return view<MEN>(edge_coordinates_)(n);  
    }else{
        // Compute the value and return 
        if(edge_coordinates_cached) return getEdgeCoordinates(n); 
        return MeshAlgorithms::getEdgeCoordinates(*this, n);
    } 
}
```
I tested it and with the `decltype(auto)` in the return type this returns a reference (no copy cstr) in the first case and a value in the second. 

In the user code this would look like: 
```
// User 
Mesh<device>* m;

Kokkos::parallel_for("",size,KOKKOS_LAMBDA(const int c){     
  // Algo or Cached
  auto&& v = m->getEdgeCoordinates(c); 
  // Cached
  auto&& v = m->getEdgeCoordinates<cached>(c); 
```
Note that we can use `auto&&` in both case also.

**Comment by ecoon** [comment]
Actually, the last of these is growing on me, partly because of the need for another access pattern -- getting things from the framework mesh.  This is only valid on host, but may still be needed in some cases.


```
template<MemSpace_type mem>
struct MeshCache {
  ...
  template<AccessPattern AP = ANY>
  decltype(auto) getNodeCoordinate(Entity_ID n) {
    if constexpr(AP == CACHED) {
      assert(node_coordinates_cached);
      return view<MEM>(node_coordinates_)(n);
    } else if constexpr(AP == FRAMEWORK && MEM == HOST) {
      assert(framework_.get());
      return framework_->getNodeCoordinate(n);
    } else if constexpr(AP == ANY) {
      if (node_coordinates_cached) return getNodeCoordinate<CACHED>(n);
      return getNodeCoordinate<FRAMEWORK>(n);
    }
    throw("Invalid access error");
  }
}
```

which looks a bit scary, or we could do (same code, just written like template specialization, this also works)...


```
template<MemSpace_type mem>
struct MeshCache {
  ...
  template<AccessPattern AP = ANY>
  decltype(auto) getNodeCoordinate(Entity_ID n) {
    if (node_coordinates_cached) return getNodeCoordinate<CACHED>(n);
    return getNodeCoordinate<FRAMEWORK>(n);
  }

  template<> inline
  decltype(auto) getNodeCoordinate<CACHED>(Entity_ID n) {
    assert(node_coordinates_cached);
    return view<MEM>(node_coordinates_)(n);
  }

  template<> inline
  decltype(auto) getNodeCoordinate<FRAMEWORK>(Entity_ID n) {
    if constexpr(mem == HOST) {
      assert(framework_.get());
      return framework_->getNodeCoordinate(n);
    } else {
      throw("Invalid access error");
    }
  }
}
```

I may be leaning toward this for a few reasons:

1. It allows us to write only specializations that make sense -- the default specialization can error.
2. Alternatively the default specialization could just be the ANY one, so if user code calls the CACHED variant but CACHED hasn't been implemented yet, it would still use the ANY one and work.  Later we add the specialization and the CACHED version magically works.
3. We can have more than two "spellings" of the function, e.g. COMPUTE, CACHED, FRAMEWORK, and ANY (default) can all coexist if that make sense for a given entity.
4. It allows us not to have to think about how the function is spelled.  While it seems easy that cached is spelled one way and not cached is spelled another, I suspect this would get annoying even if it seems natural now.

**Reply by JulienLoiseau** [reply]
@ecoon (not sure if you get notification when I just write here. I dont) 

We might not have a choice here. 
We cannot specify the member function `getNodeCoordinate` inside the class (or outside for that matter) since we cannot do full specialization of a template member in a template class. 
Aka we cannot write (same thing inside the decl header): 

```
template<MemSpace_type MEM>                  // Template of MeshCache
template<AccessPattern AP = DEFAULT>      // Template of the function getNodeCoordinate
decltype(auto) MeshCache<MEM>::getNodeCoordinate(Entity_ID n) {
    if (node_coordinates_cached) return getNodeCoordinate<CACHED>(n);
    return getNodeCoordinate<FRAMEWORK>(n);
}

template<MemSpace_type MEM>
template<> inline
decltype(auto) MeshCache<MEM>::getNodeCoordinate<CACHED>(Entity_ID n) {
    assert(node_coordinates_cached);
    return view<MEM>(data_.node_coordinates_)(n);
}
...
```
This is not allowed and errors with: 
```
explicit specialization in non-namespace scope ...
```
https://stackoverflow.com/questions/2097811/c-syntax-for-explicit-specialization-of-a-template-function-in-a-template-clas

If we want to do this have to specify the class too: 
```
// Host
template<>
template<> inline
decltype(auto) MeshCache<HOST>::getNodeCoordinate<CACHED>(Entity_ID n) {
    assert(node_coordinates_cached);
    return view<MEM>(node_coordinates_)(n);
}

// Device 
template<>
template<> inline
decltype(auto) MeshCache<DEVICE>::getNodeCoordinate<CACHED>(Entity_ID n) {
    assert(node_coordinates_cached);
    return view<MEM>(node_coordinates_)(n);
}
...

```

The version with the `if constexpr` works. 
I am looking if there is a more elegant solution.

**Reply by JulienLoiseau** [reply]
@ecoon 
I have something that would also save a lot of duplication in the code. 
In the `MeshDefs.hh` I wrote a `Getter` and `RaggerGetter` struct that we can then use for each of the callers. 

Here is the example of the `Getter` (for `DualView`)

```
// Getter for DualView
template<MemSpace_type MEM, AccessPattern AP = AccessPattern::DEFAULT>
struct Getter{ 
  template<typename DATA, typename F>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f,  const Entity_ID n){ 
    if (cached) return View<MEM>(d)[n];
    if(f) return f(n);
    throwAccessError_("getNodeCoordinate");
    return AmanziGeometry::Point();
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::CACHE>{
  template<typename DATA, typename F>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f,  const Entity_ID n){ 
    return View<MEM>(d)[n];
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::FRAMEWORK>{
  template<typename DATA, typename F>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator() (bool cached, DATA& d, F&& f,  const Entity_ID n){ 
    return f(n);
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::COMPUTE>{
  template<typename DATA, typename F>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f, const Entity_ID n){ 
    throwAccessError_("getNodeCoordinate");
    return AmanziGeometry::Point();
  }
};
```

This is then used in the `MeshCache_decl.hh` as: 

```
  template<AccessPattern AP = AccessPattern::DEFAULT>
  KOKKOS_INLINE_FUNCTION 
  decltype(auto) // AmanziGeometry::Point
  getNodeCoordinate(const Entity_ID n) const { 
    return Getter<MEM,AP>(
        node_coordinates_cached,
        data_.node_coordinates,
        framework_mesh? framework_mesh_->getNodeCoordinate:nullptr, 
        n); 
  }
```

This is a first version that we can refine. 
We can also declare these `operator()`, `KOKKOS_FORCE_INLINE_FUNCTION` to be certain it is inlined. 

Note: It is returning Point for now in the error case. This will just return an empty type of the same type as the DualView of course.

**Reply by JulienLoiseau** [reply]
Updated version handling the Compute case: 

```
// Getter for DualView
template<MemSpace_type MEM, AccessPattern AP = AccessPattern::DEFAULT>
struct Getter{ 
  template<typename DATA, typename F, typename C>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f, C&&c,  const Entity_ID n){ 
    if (cached) return View<MEM>(d)[n];
    if(f) return f(n);
    if(c) return c(n); 
    //throwAccessError_("getNodeCoordinate");
    return AmanziGeometry::Point();
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::CACHE>{
  template<typename DATA, typename F, typename C>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f, C&& c, const Entity_ID n){ 
    return View<MEM>(d)[n];
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::FRAMEWORK>{
  template<typename DATA, typename F, typename C>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator() (bool cached, DATA& d, F&& f, C&& c, const Entity_ID n){ 
    return f(n);
  }
};

template<MemSpace_type MEM>
struct Getter<MEM,AccessPattern::COMPUTE>{
  template<typename DATA, typename F, typename C>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(bool cached, DATA& d, F&& f, C&& c, const Entity_ID n){
    if(c) return c(n);  
    //throwAccessError_("getNodeCoordinate");
    return AmanziGeometry::Point();
  }
};
```

**Reply by ecoon** [reply]
@JulienLoiseau 

Ugh, that's unfortunate.  I guess the generic getter works, and agreed that it would save a ton of code if it did.  And I think it should be general enough, as long as it deals with the cases where f and/or c are nullptr correctly.  We may have to still pass in the MeshCache object as well, e.g.:


```
// Getter for DualView
template<MemSpace_type MEM, AccessPattern AP = AccessPattern::DEFAULT>
struct Getter{ 
  template<typename DATA, typename F, typename C>
  KOKKOS_INLINE_FUNCTION decltype(auto)
  operator()(const MeshCache& m, bool cached, DATA& d, F&& f, C&&c,  const Entity_ID n){ 
    if (cached) return View<MEM>(d)[n];
    if(f) return f(n);
    if(c) return c(m, n); 
    //throwAccessError_("getNodeCoordinate");
    return AmanziGeometry::Point();
  }
};
```

And pass that in as well, e.g.

```
  template<AccessPattern AP = AccessPattern::DEFAULT>
  KOKKOS_INLINE_FUNCTION 
  decltype(auto) // AmanziGeometry::Point
  getNodeCoordinate(const Entity_ID n) const { 
    return Getter<MEM,AP>(
        *this,
        node_coordinates_cached,
        data_.node_coordinates,
        framework_mesh ? framework_mesh_->getNodeCoordinate : nullptr, 
        MeshAlgorithms::getNodeCoordinate,
        n); 
  }
```

The possible downside of this is we have reintroduced a branch prediction on framework_mesh, so that even `m->getNodeCoordinate<CACHED>(n);` will require a runtime check.

But that's probably ok?  Or maybe it is ok to have one `if constexpr` to deal with that case, but deal with the rest in the Getter?


```
  template<AccessPattern AP = AccessPattern::DEFAULT>
  KOKKOS_INLINE_FUNCTION 
  decltype(auto) // AmanziGeometry::Point
  getNodeCoordinate(const Entity_ID n) const { 
    if constexpr(AP == CACHED) {
      return view<MEM>(data_.node_coordinates);
    } else {
      return Getter<MEM,AP>(
          *this,
          node_coordinates_cached,
          data_.node_coordinates,
          framework_mesh ? framework_mesh_->getNodeCoordinate : nullptr, 
          MeshAlgorithms::getNodeCoordinate,
          n);
    } 
  }
```

**Reply by JulienLoiseau** [reply]
@ecoon We can talk about this during today's meeting. 
I have some updates on my side.

**Comment by ecoon** [comment]
@JulienLoiseau 

One more wrinkle -- I'm realizing our compute might not work as planned anyway.  On device, we won't be able to allocate space.  Unless we have some designated "work space" saved in the MeshCache object, I'm not sure how to deal with this.

Consider the function(s), ignoring some boilerplate:

```
template<>
KOKKOS_INLINE_FUNCTION
Point 
MeshCache<MEM>::getCellCentroid<AccessPattern::COMPUTE>(int c) {
  auto cnode_coords = m.getCellCoordinates(c);
  // do some math on cnode_coords to compute centroid cc
  return cc;
}

template<AccessPattern AP = AccessPattern::DEFAULT>
KOKKOS_INLINE_FUNCTION
Kokkos::View<Point*, MEM> 
MeshCache<MEM>::getCellCoordinates(c) {
   auto cnodes = m.getCellNodes(c);
   Kokkos::View<Point*, MEM> cnode_coords(cnodes.extent(0));  /// <---- this cannot be on device!
   for (int i=0; i!=cnodes.extent(0); ++i) cnode_coords(i) = m.getNodeCoordinates(cnodes(i));
   return cnode_coords;
}

```

So where do we create memory for cnode_coords?

It gets worse.

```
template<AccessPattern AP = AccessPattern::DEFAULT>
KOKKOS_INLINE_FUNCTION
Kokkos::View<int*, MEM> 
MeshCache<MEM>::getCellNodes(c) {
   // this method simply forms the union of all cell-face-nodes
   Kokkos::View<Point*, MEM> cnodes(MAX_NUM_NODES);  /// <---- this cannot be on device!
   auto cfaces = m.getCellFaces(c);
   int index = 0;
   for (int i=0; i!=cfaces.extent(0); ++i) {
      auto fnodes = m.getFaceNodes(f);
      // pseudocode...
      for (auto n : fnodes) if (n ! in cnodes) cnodes(index++) = n;
   }
   cnodes.resize(index);
   return cnodes;
}
```

So really, to implement this compute method `getCellCentroid()` which looks like it involves no views requires two separate "workspace" views, one for coordinates and one for indices, both of which we somehow need to have preallocated?
